File: packages/core/src/domain/CostEstimator.ts
import type { ModelPricing } from '../ai/PricingFetcher.js';

const AVG_TOKENS_PER_ACTION = 2_500;
const AVG_TOKENS_GENERATION = 8_000;

export interface CostEstimate {
  exploration: number;
  generation: number;
  total: number;
}

export function estimateStartupCost(
  estimatedActions: number,
  prices: { fast: ModelPricing | undefined; strong: ModelPricing | undefined },
): CostEstimate {
  const exploration = prices.fast
    ? estimatedActions * AVG_TOKENS_PER_ACTION * prices.fast.inputUsdPerToken
    : 0;

  const generation = prices.strong
    ? AVG_TOKENS_GENERATION * prices.strong.inputUsdPerToken
    : 0;

  return {
    exploration,
    generation,
    total: exploration + generation,
  };
}

File: packages/core/python/src/application/use_cases/startup/cost_estimator.py
from __future__ import annotations

from dataclasses import dataclass

from src.application.use_cases.startup.token_budgets import (
    AVG_TOKENS_GENERATION,
    AVG_TOKENS_PER_ACTION,
)
from src.domain.model.model_price import ModelPrice


@dataclass
class CostEstimate:
    exploration: float
    generation: float
    total: float


def estimate_cost(
    estimated_actions: int,
    model_fast_price: ModelPrice | None,
    model_strong_price: ModelPrice | None,
) -> CostEstimate:
    """Estimate cost in USD for an exploration + generation session.

    Args:
        estimated_actions: Number of exploration rounds expected.
        model_fast_price: Pricing for the fast model (exploration phase).
            If None, exploration cost is 0.0.
        model_strong_price: Pricing for the strong model (generation phase).
            If None, generation cost is 0.0.

    Returns:
        CostEstimate with exploration, generation and total costs in USD.
    """
    if model_fast_price is not None:
        exploration = (
            estimated_actions
            * AVG_TOKENS_PER_ACTION
            * model_fast_price.input_usd_per_token
        )
    else:
        exploration = 0.0

    if model_strong_price is not None:
        generation = AVG_TOKENS_GENERATION * model_strong_price.input_usd_per_token
    else:
        generation = 0.0

    total = exploration + generation

    return CostEstimate(exploration=exploration, generation=generation, total=total)
